### ISSUE TITLE
Strange behavior with stdout/stderr from Python

### ISSUE THREAD

**Issue by estan** [thp/pyotherside#61]
Sorry for the slightly vague issue title, but I'm hoping to get some clarification.

With the test case

`main.qml`:

```
import QtQuick 2.0
import io.thp.pyotherside 1.4

Rectangle {
    color: 'black'
    width: 400
    height: 400

    Python {
        id: py

        Component.onCompleted: {
            addImportPath(Qt.resolvedUrl('.'));

            importModule('main', function () {
                py.call('main.testprint', [], function(result) {
                    console.log('printed from QML');
                });
            });
        }
    }
}
```

`main.py`:

```
def testprint():
    print('printed from python')
```

I get this behavior:

```
estan@newton:~/printtest$ qmlscene main.qml
printed from python
qml: printed from QML
```

as expected. The `console.log` message from QML is printed on stderr (since results in a `qDebug()`). And the `print` from Python shows up as well, presumably on `stdout` (?).

However, if I do

```
estan@newton:~/printtest$ qmlscene main.qml > test.out
qml: printed from QML
estan@newton:~/printtest$ cat test.out
```

you can see that I seem unable to redirect the stdout from the Python process into a file. I guess this has something to do with how pyotherside starts the Python process?

I think related to this is the fact that running the test case from inside Qt Creator, I can't see the Python printouts, because the redirection that Qt Creator sets up can't "see" them.

Could the way pyotherside starts the Python process possibly be changed, so that the two processes share file descriptors? Or is there some other way to make this more intuitive? It's a little awkward to have to set up logging to a file in order to even see the Python output when working from Qt Creator.

**Comment by estan** [comment]
I realize now that I've slightly misunderstood in that pyotherside starts a thread for the interpreter, not a process. But I'd still like to get all output (QML and Python) to stdout/stderr file descriptors that I can redirect in the shell, and which Qt Creator can access.

**Comment by thp** [comment]
Console log from QML might go to stderr, can you try the following?

```
qmlscene main.qml >test.out 2>&1
```

**Comment by estan** [comment]
Yep, I know it goes to stderr, like I said:

> The console.log message from QML is printed on stderr (since results in a qDebug())").

What puzzles me is

```
estan@newton:~/printtest$ qmlscene main.qml > test.out
qml: printed from QML
estan@newton:~/printtest$ cat test.out
```

Here, the QML output is shown on stderr, as expected. And the output from Python is suppressed since I'm redirecting stdout to `test.out`. But the "printed from python" does not end up `test.out`, the file is empty. I made this test case because I wasn't getting the Python output in Qt Creator. I think this is related somehow.

**Comment by thp** [comment]
Might be some buffering issues. Have you tried for tests to have `sys.stdout.flush()` (with the corresponding `import`) after the `print()` call? Could be that output is buffered slightly different if redirected to a file, and the buffers can't be flushed early enough.

**Comment by estan** [comment]
Bingo that was it, thanks!

Is there any way to get this to work out of the box without having to flush stdout at every `print` call, or printing to stderr from Python? It's very convenient to use Qt Creator when editing the QML parts of the application, because of its auto-completion, and fire it up from there, but at the same time very convenient to do quick debugging using `print` from the Python parts. At the moment the plain `print`s do not show up in the Qt Creator output window due to this buffering.

**Comment by estan** [comment]
There's a `PYTHONUNBUFFERED` environment variable that can be set, but I tried it and there was no change in behavior (I verified that it was seen in `main.py`).

**Comment by estan** [comment]
I guess that doesn't work because that's something the Python checks for in it's command line interface. Maybe pyotherside could support a similar variable and do the necessary configuration of its interpreter?

**Comment by estan** [comment]
I realized this is probably not something pyotherside should do, but that applications should do themselves. What I did was

```python
sys.stdout = os.fdopen(sys.stdout.fileno(), 'w', 1)
```

if `UNBUFFERED_STDOUT` is set, to set the buffer size to 1 (Python 3 doesn't allow 0 for text streams), and then make sure I set that when running in Qt Creator. For some reason it wouldn't work when tried the same using `setbuf` in my C++ launcher application (perhaps it's not the same file descriptors...?).

Anyway, closing this issue now, thanks for the input!
